### PR TITLE
Fix race competition when the breakpoint needs be evaluated in multi threads

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

Fix microsoft/vscode-java-debug#782

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java [hunk 1]
```diff
@@ -155,7 +168,7 @@ public Object getProperty(Object key) {
         requests.add(classPrepareRequest);
 
         CompletableFuture<IWatchpoint> future = new CompletableFuture<>();
-        Disposable subscription = eventHub.events()
+        subscription = eventHub.events()
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java:171]: add subscriptions.add(subscription)?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java [hunk 2]
```diff
@@ -249,4 +262,14 @@ public void setCompiledLogpointExpression(Object compiledExpression) {
...
     }
+
+    @Override
+    public Object getCompiledExpression(long threadId) {
+        return compiledExpressions.get(threadId);
+    }
+
+    @Override
+    public void setCompiledExpression(long threadId, Object compiledExpression) {
+        compiledExpressions.put(threadId, compiledExpression);
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java:273]: put/get in multiple thread is error prone.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java [hunk 3]
```diff
@@ -148,20 +148,12 @@ public void initialize(IDebugAdapterContext context, Map<String, Object> props)
             ASTEvaluationEngine engine = new ASTEvaluationEngine(project, debugTarget);
             boolean newExpression = false;
             if (breakpoint != null) {
-                if (StringUtils.isNotBlank(breakpoint.getLogMessage())) {
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java:151]: why this condition is deleted: StringUtils.isNotBlank(breakpoint.getLogMessage())?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java (hunk 1 above)
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java:171]: the following L182 already did that.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java (hunk 3 above)
Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java:151]: this is not needed any more. Previously the compiled expression for the logpoint and conditional breakpoint are saved separately, so when evaluating a breakpoint, need check whether it's a logpoint or conditional breakpoint. In the new implementation, they are saved in a map, and fetched from the map directly.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java (hunk 2 above)
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/Watchpoint.java:273]: nice catch!

**Pull request merged**